openedx_tagging/core/tagging/models/system_defined.py

```diff
@@ -0,0 +1,271 @@
+
+    Tag.external_id stores an identifier from the instance (`pk` as default)
+    and Tag.value stores a human readable representation of the instance
+    (e.g. `username`).
+    The subclasses can override this behavior, to choose the right field.
+
+    When an ObjectTag is created with an existing Tag,
+    the Tag is re-synchronized with its instance.
+    """
+
+    class Meta:
+        proxy = True
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """
+        Checks if the `object_tag_class` is a subclass of ModelObjectTag.
+        """
+        assert issubclass(self.object_tag_class, ModelObjectTag)
+        super().__init__(*args, **kwargs)
+
+    @property
+    def object_tag_class(self) -> Type:
+        """
+        Returns the ObjectTag subclass associated with this taxonomy.
+
+        Model Taxonomy subclasses must implement this to provide a ModelObjectTag subclass.
+        """
+        raise NotImplementedError
+    
+    def _check_instance(self, object_tag: ObjectTag) -> bool:
+        """
+        Returns True if the instance exists
+
+        Subclasses can override this method to perform their own instance validation checks.
+        """
+        object_tag = self.object_tag_class.cast(object_tag)
+        return bool(object_tag.get_instance())        
+
+    def _check_tag(self, object_tag: ObjectTag) -> bool:
+        """
+        Returns True if the instance is valid
+        """
+        return super()._check_tag(object_tag) and self._check_instance(object_tag)
+
+
+class UserModelObjectTag(ModelObjectTag):
+    """
+    ObjectTags for the UserSystemDefinedTaxonomy.
+    """
+
+    class Meta:
+        proxy = True
+
+    @property
+    def tag_class_model(self) -> Type:
+        """
+        Associate the user model
+        """
+        return get_user_model()
+
+    @property
+    def tag_class_value(self) -> str:
+        """
+        Returns the name of the tag_class_model field to use as the Tag.value when creating Tags for this taxonomy.
+
+        Subclasses may override this method to use different fields.
+        """
+        return "username"
+
+
+class UserSystemDefinedTaxonomy(ModelSystemDefinedTaxonomy):
+    """
+    User based system taxonomy class.
+    """
+
+    class Meta:
+        proxy = True
+
+    @property
+    def object_tag_class(self) -> Type:
+        """
+        Returns the ObjectTag subclass associated with this taxonomy, which is ModelObjectTag by default.
+
+        Model Taxonomy subclasses must implement this to provide a ModelObjectTag subclass.
+        """
+        return UserModelObjectTag
+
+
+class LanguageTaxonomy(SystemDefinedTaxonomy):
+    """
+    Language System-defined taxonomy
+
+    The tags are filtered and validated taking into account the
+    languages available in Django LANGUAGES settings var
+    """
+
+    class Meta:
+        proxy = True
+
+    def get_tags(self, tag_set: models.QuerySet = None) -> List[Tag]:
```

Comment:
nit: This `tag_set` parameter is unused here, and undocumented in the base class. I think the API is cleaner if you just keep it as `get_tags(self) -> List[Tag]` and move the prefetch/sort code to a helper method like `Taxonomy._sort_tag_results()`. Up to you though.

Reply:
Funny.. that's exactly what @ChrisChV did initially, and I asked him to change it to this instead, because I think this is cleaner :)

I'll note this in my cleanup task, but to keep things from flip-flopping here, I'd rather leave it as-is for now.

Thank you @bradenmacdonald !

Reply:
haha ooops... I didn't read the earlier review. And this is definitely subjective.